gcsio/src/main/java/com/google/cloud/hadoop/gcsio/DeleteFolderOperation.java

```diff
@@ -0,0 +1,257 @@
+   * queue
+   *
+   * @param folderResource of the folder that is now deleted
+   */
+  private void successfullDeletionOfFolderResource(FolderInfo folderResource) {
+    // remove the folderResource from list of map
+    this.countOfChildren.remove(folderResource.getFolderName());
+
+    String parentFolder = folderResource.getParentFolderName();
+    if (this.countOfChildren.containsKey(parentFolder)) {
+
+      // update the parent's count of children
+      this.countOfChildren.replace(parentFolder, this.countOfChildren.get(parentFolder) - 1);
+
+      // if the parent folder is now empty, append in the queue
+      if (this.countOfChildren.get(parentFolder) == 0) {
+        addFolderResourceInBlockingQueue(
+            new FolderInfo(
+                FolderInfo.createFolderInfoObject(folderResource.getBucket(), parentFolder)));
+      }
+    }
+  }
+
+  /** Helper function to delete a single folder resource */
+  private void queueSingleFolderDelete(@Nonnull final FolderInfo folder, final int attempt) {
+    final String bucketName = folder.getBucket();
+    final String folderName = folder.getFolderName();
+    checkArgument(
+        !Strings.isNullOrEmpty(bucketName),
+        String.format("No bucket for folder resource %s", bucketName));
+    checkArgument(
+        !Strings.isNullOrEmpty(folderName),
+        String.format("No folder path for folder resource %s", folderName));
+
+    addToToBatchExecutorQueue(
+        new DeleteFolderRequestCallable(folder, this.storageControlClient), folder, attempt);
+  }
+
+  /** Helper to create a callback for a particular deletion request for folder. */
+  private FutureCallback getDeletionCallback(
+      final FolderInfo resourceId,
+      final KeySetView<IOException, Boolean> allExceptions,
+      final int attempt) {
+    return new FutureCallback<Void>() {
+      @Override
+      public void onSuccess(Void result) {
+        logger.atFiner().log("Successfully deleted folder %s", resourceId.toString());
+        successfullDeletionOfFolderResource(resourceId);
+      }
+
+      @Override
+      public void onFailure(Throwable throwable) {
+        if (throwable instanceof Exception
+            & errorTypeExtractor.getErrorType((Exception) throwable) == ErrorType.NOT_FOUND) {
+          // Ignore item-not-found errors. We do not have to delete what we cannot find.
+          // This
+          // error typically shows up when we make a request to delete something and the
+          // server
+          // receives the request but we get a retry-able error before we get a response.
+          // During a retry, we no longer find the item because the server had deleted
+          // it already.
+          logger.atFiner().log(
+              "Delete folder '%s' not found: %s", resourceId, throwable.getMessage());
+          successfullDeletionOfFolderResource(resourceId);
+        } else if (throwable instanceof Exception
```

Comment:
nit: refactoring this to a function will be more readable

Reply:
Done